### PR TITLE
Internal: Editor V2 - Add edit mode listener [ED-9655]

### PR DESCRIPTION
**Pull request by StyleShit**

### CODE REVIEW

File: packages/packages/v1-adapters/src/dispatchers/dispatchers.ts [hunk 1]
```diff
@@ -37,8 +37,14 @@ export function openRoute( route: string ): Promise<void> {
...
 	const extendedWindow = window as unknown as ExtendedWindow;
 
 	return !! extendedWindow.$e?.routes?.isPartOf( route );
 }
+
+export function getCurrentEditMode() {
+	const extendedWindow = window as unknown as ExtendedWindow;
+
+	return extendedWindow.elementor.channels.dataEditMode.request( 'activeMode' );
+}
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/dispatchers/dispatchers.ts:50]: Why those 2 related to dispatchers?

File: packages/packages/v1-adapters/src/dispatchers/dispatchers.ts [hunk 2]
```diff
@@ -37,8 +37,14 @@ export function openRoute( route: string ): Promise<void> {
...
+export function isRouteActive( route: string ) {
 	const extendedWindow = window as unknown as ExtendedWindow;
 
 	return !! extendedWindow.$e?.routes?.isPartOf( route );
 }
+
+export function getCurrentEditMode() {
+	const extendedWindow = window as unknown as ExtendedWindow;
+
+	return extendedWindow.elementor.channels.dataEditMode.request( 'activeMode' );
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/dispatchers/dispatchers.ts:49]:
```suggestion
	return extendedWindow?.elementor?.channels?.dataEditMode?.request( 'activeMode' );
```

to support using the code without the editor v1

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: packages/packages/v1-adapters/src/dispatchers/dispatchers.ts (hunk 1 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/dispatchers/dispatchers.ts:50]: actually they're not really related, just didn't know where to put them, so I counted on the fact that you'll catch that in the review 😂

File: packages/packages/v1-adapters/src/dispatchers/dispatchers.ts [hunk 3]
```diff
@@ -37,8 +37,14 @@ export function openRoute( route: string ): Promise<void> {
 	}
 }
 
-export function isRouteActive( route: string ): boolean {
+export function isRouteActive( route: string ) {
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/dispatchers/dispatchers.ts:40]: move to utils.

File: packages/packages/v1-adapters/src/hooks/__tests__/use-is-preview-mode.test.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+		const { result } = renderHook( () => useIsPreviewMode() );
+
+		// Act.
+		act( () => {
+			mockedGetCurrentEditMode.mockReturnValue( 'preview' );
+			dispatchEditModeChange();
+		} );
+
+		// Assert.
+		expect( result.current ).toBe( true );
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/hooks/__tests__/use-is-preview-mode.test.ts:40]: First, verify that it's `false`.

**Review by rotemee — CHANGES_REQUESTED** [review]

File: packages/packages/elements-panel/src/hooks/use-props.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import * as React from 'react';
+import { __ } from '@wordpress/i18n';
+import { SvgIcon } from '@elementor/ui';
+import { openRoute, useIsPreviewMode, useIsRouteActive } from '@elementor/v1-adapters';
+
+export default function useProps() {
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/packages/elements-panel/src/hooks/use-props.tsx:6]: Done

File: packages/packages/elements-panel/src/hooks/use-props.tsx [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	const isInElementsPanel = useIsRouteActive( 'panel/elements' );
+	const isInSiteSettings = useIsRouteActive( 'panel/global' );
+	const isPreview = useIsPreviewMode();
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/packages/elements-panel/src/hooks/use-props.tsx:9]: Done

File: packages/packages/elements-panel/src/hooks/use-props.tsx [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+	const selected = isInElementsPanel && ! isPreview;
+	const disabled = isPreview || isInSiteSettings;
+
+	return {
+		title: __( 'Add element', 'elementor' ),
+		icon: () => (
+			<SvgIcon viewBox="0 0 24 24">
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/packages/elements-panel/src/hooks/use-props.tsx:17]: Done

File: packages/packages/elements-panel/src/hooks/__tests__/use-props.test.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+	afterEach( () => {
+		jest.clearAllMocks();
+	} );
+
+	it( 'should open the elements panel on click', () => {
+		// Arrange.
+		const { result } = renderHook( () => useProps() );
+
+		// Act.
+		result.current.onClick?.();
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/packages/elements-panel/src/hooks/__tests__/use-props.test.tsx:24]: Done

File: packages/packages/v1-adapters/src/dispatchers/dispatchers.ts (hunk 3 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/dispatchers/dispatchers.ts:40]: Done

File: packages/packages/v1-adapters/src/hooks/__tests__/use-is-preview-mode.test.ts (hunk 4 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/hooks/__tests__/use-is-preview-mode.test.ts:40]: Done

**Review by rotemee — DISMISSED** [review]

File: packages/packages/v1-adapters/src/hooks/use-is-preview-mode.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+		const cleanup = listenTo(
+			editModeChangeEvent(),
+			updateState,
+		);
+
+		return cleanup;
+	}, [] );
+
+	return isPreview;
+}
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/hooks/use-is-preview-mode.ts:19]:
```suggestion
export default function useListenTo( events, checkState ) {
	const [ state, setState ] = useState( checkState );

	useEffect( () => {
		const updateState = () => setState( checkState() );

		const cleanup = listenTo(
			events,
			updateState,
		);

		return cleanup;
	}, [] );

	return state;
}

useListenTo( editModeChangeEvent(), () => getCurrentEditMode() === 'preview' )
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: packages/packages/v1-adapters/src/hooks/use-is-preview-mode.ts (hunk 9 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/hooks/use-is-preview-mode.ts:19]: Great idea!
changed it a little bit, WDYT?

File: assets/dev/js/editor/editor-base.js [hunk 10]
```diff
@@ -1210,6 +1210,12 @@ export default class EditorBase extends Marionette.Application {
 	onEditModeSwitched() {
 		const activeMode = this.channels.dataEditMode.request( 'activeMode' );
 
+		dispatchEvent( new CustomEvent( 'elementor/v1/edit-mode/change', {
```
Comment (Nevoss) [inline assets/dev/js/editor/editor-base.js:1213]: Talk with @rami-elementor

File: packages/packages/v1-adapters/src/index.ts [hunk 11]
```diff
@@ -1,3 +1,4 @@
 export * from './dispatchers';
 export * from './hooks';
 export * from './listeners';
+export * from './utils';
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/index.ts:4]:
```suggestion
export * from './readers';
```

File: packages/packages/v1-adapters/src/utils/__tests__/index.test.ts [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+	elementor: {
+		channels: {
+			dataEditMode: {
+				request: jest.Mock;
+			}
+		}
+	}
+}
+
+describe( '@elementor/v1-adapters - Utils', () => {
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/utils/__tests__/index.test.ts:18]:
```suggestion
describe( '@elementor/v1-adapters - Readers', () => {
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: packages/packages/v1-adapters/src/index.ts (hunk 11 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/index.ts]: Done

File: packages/packages/v1-adapters/src/utils/__tests__/index.test.ts (hunk 12 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/utils/__tests__/index.test.ts]: Done

**Review by Nevoss — DISMISSED** [review]

File: assets/dev/js/editor/editor-base.js (hunk 10 above)
Comment (rami-elementor) [reply assets/dev/js/editor/editor-base.js:1213]: According to the current convention it should be `elementor/edit-mode/change`.

Comment (rami-elementor) [reply assets/dev/js/editor/editor-base.js:1213]: Please let's not change the conventions again. We have many inconsistencies, we don't need another one.

**Review by Nevoss — APPROVED** [review]

**Pull request merged**